galsim/phase_psf.py

```diff
@@ -0,0 +1,1017 @@
+#
+"""@file phase_psf.py
+Utilities for creating PSFs from phase screens.  Essentially evaluates the Fourier optics
+diffraction equation:
+
+PSF(x, y) = int( |FT(aperture(u, v) * exp(i * phase(u, v, x, y, t)))|^2, dt)
+
+where x, y are focal plane coordinates and u, v are pupil plane coordinates
+
+The main classes of note are:
+
+FrozenAtmosphericScreen
+  Class implementing phase(u, v, x, y, t) for frozen flow von Karman type turbulence
+
+ARAtmosphericScreen
+  Class implementing phase(u, v, x, y, t) for "boiling" von Karman type turbulence
+
+PhaseScreenList
+  Python sequence type to hold multiple phase screens, for instance to simulate turbulence at
+  different altitudes.  A key method is getPSF(), which will take the list of phase screens,
+  add them together linearly (Fraunhofer approximation), and evaluate the above diffraction
+  equation to yield a PhaseScreenPSF object.
+
+PhaseScreenPSF
+  A GSObject holding the evaluated PSF.
+
+Atmosphere
+  Convenience function to quickly assemble multiple AtmosphericScreens into a PhaseScreenList.
+"""
+
+import copy
+import numpy as np
+import galsim
+import utilities
+from galsim import GSObject
+
+
+def listify(arg):
+    """Turn argument into a list if not already iterable."""
+    return [arg] if not hasattr(arg, '__iter__') else arg
+
+
+def broadcast(arg, n):
+    """Turn length-1 iterable into length-n list."""
+    return [arg[0]]*n if len(arg) == 1 else arg
+
+
+def generate_pupil(nu, pupil_size, diam=None, obscuration=None):
+    """ Generate a pupil transmission array (0's and 1's) for a circular aperture and potentially a
+    central circular obscuration.
+
+    @param nu          Number of pixels across pupil array.
+    @param pupil_size  Physical size of pupil array in meters.
+    @param diam        Diameter of aperture in meters.
+    @param obscuration Fractional linear obscuration.
+    @returns array of 0's and 1's indicating pupil transmission function.
+    """
+    aper = np.ones((nu, nu), dtype=np.float64)
+    if diam is not None:
+        radius = 0.5*diam
+        u = np.fft.fftshift(np.fft.fftfreq(nu, 1./pupil_size))
+        u, v = np.meshgrid(u, u)
+        rsqr = u**2 + v**2
+        aper = rsqr <= radius**2
+        if obscuration is not None:
+            aper *= rsqr >= (radius*obscuration)**2
+    return aper
+
+
+class PhaseScreen(object):
+    # ABC for phase screens.  Subclasses should implement:
+    # advance()
+    # advance_by()
+    # reset()
+    # path_difference()
+    # pupil_scale()
+    # Non-evolving screens, such as one representing optical aberrations, can probably accept the
+    # default advance(), advance_by(), and reset() methods which are all no-ops.  All subclasses
+    # will need to implement their own path_difference() and pupil_scale() though.
+    """ Abstract base class for a phase screen to use in generating a PSF using Fourier optics.
+    Not intended to be instantiated directly.
```

Comment:
>I don't see the point of this base class. There don't seem to be any sub-classes of it other than AtmosphericScreen. Is there a plan to extend this in a different direction so eventually there will be another subclass of this, where this abstraction might be relevant?

A few updates.  I just finished adding an `OpticalScreen` class, which mimics the functionality in optics.py.  (Though with an extra feature: aberrations can now go to arbitrarily high Zernikes!)  

After looking at the result and reading through (http://stackoverflow.com/questions/3570796/why-use-abstract-base-classes-in-python) I decided to just use duck-typing instead of an ABC and deleted the `PhaseScreen` ABC entirely.  This does raise a question related to style though:

`OpticalScreen` is time independent, which means there's no reason to define methods `advance`, `advance_by`, or `reset` that are present in `AtmosphericScreen`.  This needs to be handled somehow by `PhaseScreenList.advance`, `advance_by`, and `reset` methods though.  I could think of three options:

- Define `OpticalScreen.advance()` as a noop, either directly or via subclassing (this was the former approach I just deleted).
- Use a `hasattr(screen, 'advance')` inside `PhaseScreenList.advance()`.
- Use a try/except in `PhaseScreenList`.  I.e., [EAFP](https://docs.python.org/2/glossary.html#term-eafp)

The latter seemed the most pythonic, so I went with that.

There's also a new `Aperture` class that handles obscurations and struts similar to `optics.py:generate_pupil_plane()`.  I haven't reimplemented `optics.py:load_pupil_plane()` to load pupil planes from images yet, but that's on the todo list.

I didn't bother with the other stuff in `optics.py`, like the `otf`, `mtf`, `ptf`, and so on.  These aren't imported into the galsim namespace by `__init__.py`, so I'm not sure if these really need to be reimplemented in the new framework or not.

And finally, I'm considering renaming the `path_difference` method to `wavefront`, which is probably a bit more standard.  It does conflict conceptually with the use of `wavefront` in `optics.py`, however.  In `phase_psf.py`, I would want `wavefront` to represent something like the number of nanometers that the physical wavefront leads or lags the ideal plane/spherical wave.  In `optics.py`, however, `wavefront` already includes the next steps needed to create the PSF, namely division by the wavelength and exponentiation.  Roughly: `optics_py_wavefront = np.exp(2j * np.pi * joshs_preferred_definition_of_wavefront / wavelength)`.

Any comments regarding implementing MTF, PTF, OTF, etc. and renaming `path_difference` to `wavefront` would be welcome!